pilot/pkg/proxy/envoy/v2/mesh_test.go

```diff
@@ -0,0 +1,765 @@
+	"math"
+	"reflect"
+	"sort"
+	"strconv"
+	"strings"
+	"testing"
+
+	route "istio.io/api/routing/v1alpha2"
+)
+
+const (
+	testService1 = "test-service-1"
+	testService2 = "test-service-2"
+	testLabelApp = "app"
+	testLabelVer = "version"
+	testLabelEnv = "environment"
+	testLabelExp = "experiment"
+	testLabelShd = "shard"
+	testSubset1  = "test-subset-1"
+	testSubset2  = "test-subset-2"
+)
+
+const (
+	_                         = iota
+	noAttrChange attrToChange = 1 << iota
+	diffName
+	diffService
+	diffNamespace
+	diffDomains
+	diffAddr
+	diffPort
+	diffProtocol
+	diffLabels
+	diffUser
+	diffServiceOrder
+	diffDomainOrder
+)
+
+var (
+	allAttrToChange = []attrToChange{
+		noAttrChange,
+		diffName,
+		diffService,
+		diffNamespace,
+		diffDomains,
+		diffAddr,
+		diffPort,
+		diffProtocol,
+		diffLabels,
+		diffUser,
+		diffServiceOrder,
+		diffDomainOrder,
+	}
+	allTestLabels = []string{
+		testLabelApp,
+		testLabelVer,
+		testLabelEnv,
+		testLabelExp,
+		testLabelShd,
+	}
+
+	// Test domain sets to associate with a service.
+	testDomainSets = [][]string{
+		{"default.domain-1.com", "my-namespace.my-own-domain-1.com", "my-other-ns.domain-1.com"},
+		{"domain-2.com", "my-own-domain-2.com"},
+		{"domain-3.com", "my-own-domain-3.com", "my-other-domain-3.com"},
+	}
+
+	// Applies only to services using testDomainSets[0]
+	testNamespaces = []string{"default", "my-namespace", "my-other-ns"}
+
+	testPorts = []uint32{80, 443, 8080}
+
+	testProtocols = []string{"http", "https", "grpc", "redis", "mongo"}
+)
+
+// Enums used to tweak attributes while
+// building an endpoint for testing
+type attrToChange int
+
+type samplePoint struct {
+	name      string
+	rule      int
+	subset    int
+	endpoints int
+}
+
+type dataSet struct {
+	cntEps           int
+	cntSvcs          int
+	cntSubsetsPerSvc int
+}
+
+func TestMeshNewMesh(t *testing.T) {
+	tm := NewMesh()
+	if tm == nil {
+		t.Error("expecting valid Mesh, found nil")
+	}
+}
+
+func TestMeshXDS(t *testing.T) {
+	tm := NewMesh()
+	expectedEps := []*Endpoint{}
+	t.Run("EmptyMesh", func(t *testing.T) {
+		actual := tm.SubsetEndpoints([]string{"test-service-1.default.domain-1.com"})
+		assertEqualEndpointLists(t, expectedEps, actual)
+	})
+	countEps, countSvcs, countSubsetsPerSvc := 32, 2, 2
+	rules, subsets, expectedEps, _ := buildTestEndpoints(t, countEps, countSvcs, countSubsetsPerSvc)
+	expectedSubsets := []string{}
+	for i := 0; i < countSvcs; i++ {
+		servicePrefix := "test-service-" + strconv.Itoa(i+1)
+		for _, domain := range testDomainSets[i] {
+			expectedSubsets = append(expectedSubsets, servicePrefix+"."+domain)
+		}
+	}
+	err := tm.Reconcile(expectedEps)
+	if err != nil {
+		t.Errorf("unable to reconcile(): %v", err)
+		return
+	}
+	t.Run("AfterReconcile", func(t *testing.T) {
+		if err != nil {
+			t.Error(err)
+			return
+		}
+		t.Run("SubsetNames", func(t *testing.T) {
+			t.Parallel()
+			actualSubsets := tm.SubsetNames()
+			assertEqualsSubsetNames(t, expectedSubsets, actualSubsets)
```

Comment:
Aside: strong argument for gomega here, I think (cc @zachgersh, would these be improved by adopting gomega?)

Reply:
Yes, pretty easy win immediately as you'd get to just delete that custom matcher. `ConsistOf()` from gomega does exactly the same thing and should provide a helpful error message when the slice is missing an element.

I'd love to see this switched over. More info about ConsistOf [here](http://onsi.github.io/gomega/#consistofelement-interface)

Pretty sure we could also dump `assertEqualEndpointLists()` in favor a simplified set of `Equal()` assertions. Could potentially use [gstruct](http://onsi.github.io/gomega/#gstruct-testing-complex-data-types) if we want to keep that slice of `*Endpoints` but I always tend to favor the simplest assertions first. @gnirodi let me know if you want more details on how to incorporate `gomega` - I am free to help.

Reply:
I've not used gomega. Could we take this up in a subsequent PR?